File: paw_acp_sdk/paw_acp_sdk/conversation.py
"""Per-session conversation history manager."""

import time
from typing import Dict, List


class ConversationManager:
    """Manages per-session conversation history with automatic trimming.

    Usage::

        conv = ConversationManager(max_history=20)
        conv.add_user_message("session_1", "Hello")
        conv.add_assistant_message("session_1", "Hi there!")
        messages = conv.get_messages("session_1")
    """

    def __init__(self, max_history: int = 20):
        self.max_history = max_history
        self._sessions: Dict[str, List[Dict[str, str]]] = {}
        self._last_access: Dict[str, float] = {}

    def get_messages(self, session_id: str) -> List[Dict[str, str]]:
        """Return the message list for a session (creates empty if needed)."""
        self._last_access[session_id] = time.time()
        return self._sessions.get(session_id, [])

    def add_user_message(self, session_id: str, content: str):
        self._ensure_session(session_id)
        self._sessions[session_id].append({"role": "user", "content": content})
        self._trim(session_id)

    def add_assistant_message(self, session_id: str, content: str):
        self._ensure_session(session_id)
        self._sessions[session_id].append({"role": "assistant", "content": content})
        self._trim(session_id)

    def rollback(self, session_id: str) -> bool:
        """Remove the last assistant+user message pair. Returns True if something was removed."""
        msgs = self._sessions.get(session_id, [])
        if not msgs:
            return False
        if msgs and msgs[-1]["role"] == "assistant":
            msgs.pop()
        if msgs and msgs[-1]["role"] == "user":
            msgs.pop()
        return True

    def has_session(self, session_id: str) -> bool:
        return session_id in self._sessions

    def initialize_session(self, session_id: str, history: List[Dict[str, str]]):
        """Pre-load a session with existing history. Only if session doesn't exist."""
        if session_id in self._sessions:
            return
        self._sessions[session_id] = list(history)
        self._last_access[session_id] = time.time()

    def prepend_history(self, session_id: str, older_messages: List[Dict[str, str]]):
        """Prepend older history messages to the beginning of an existing session."""
        if session_id not in self._sessions:
            return
        self._sessions[session_id] = older_messages + self._sessions[session_id]
        self._last_access[session_id] = time.time()

    def cleanup_expired(self, max_age_seconds: int = 259200):
        """Remove sessions older than max_age_seconds."""
        now = time.time()
        expired = [sid for sid, ts in self._last_access.items()
                   if now - ts > max_age_seconds]
        for sid in expired:
            self._sessions.pop(sid, None)
            self._last_access.pop(sid, None)

    def _ensure_session(self, session_id: str):
        if session_id not in self._sessions:
            self._sessions[session_id] = []
        self._last_access[session_id] = time.time()

    def _trim(self, session_id: str):
        msgs = self._sessions[session_id]
        max_msgs = self.max_history * 2
        if len(msgs) > max_msgs:
            self._sessions[session_id] = msgs[-max_msgs:]

File: sdks/shepaw-acp-sdk-typescript/src/conversation.ts
/**
 * Per-session conversation history manager.
 *
 * Wire-compatible with `shepaw_acp_sdk.conversation.ConversationManager`:
 *   - Per-session `messages[]` with automatic trimming to `maxHistory` turns
 *     (each turn = 2 messages: user + assistant, so cap is `maxHistory * 2`).
 *   - Per-session last-access timestamp for TTL-based cleanup.
 *   - `rollback()` removes the trailing assistant + user pair (if present).
 */

import type { ConversationMessage } from './types.js';

export interface ConversationManagerOptions {
  /** Maximum number of turns (user+assistant pairs) to keep per session. Default 20. */
  maxHistory?: number;
}

export class ConversationManager {
  readonly maxHistory: number;

  private readonly sessions = new Map<string, ConversationMessage[]>();
  private readonly lastAccess = new Map<string, number>();

  constructor(opts: ConversationManagerOptions = {}) {
    this.maxHistory = opts.maxHistory ?? 20;
  }

  /** Return the message list for a session (empty array if it doesn't exist). */
  getMessages(sessionId: string): ConversationMessage[] {
    this.touch(sessionId);
    return this.sessions.get(sessionId) ?? [];
  }

  addUserMessage(sessionId: string, content: string): void {
    this.ensureSession(sessionId);
    this.sessions.get(sessionId)!.push({ role: 'user', content });
    this.trim(sessionId);
  }

  addAssistantMessage(sessionId: string, content: string): void {
    this.ensureSession(sessionId);
    this.sessions.get(sessionId)!.push({ role: 'assistant', content });
    this.trim(sessionId);
  }

  /**
   * Remove the last assistant+user pair.
   * Returns `true` when at least one message was removed.
   */
  rollback(sessionId: string): boolean {
    const msgs = this.sessions.get(sessionId);
    if (!msgs || msgs.length === 0) return false;
    if (msgs.at(-1)?.role === 'assistant') msgs.pop();
    if (msgs.at(-1)?.role === 'user') msgs.pop();
    return true;
  }

  hasSession(sessionId: string): boolean {
    return this.sessions.has(sessionId);
  }

  /** Pre-load a session with existing history. No-op if the session already exists. */
  initializeSession(sessionId: string, history: ConversationMessage[]): void {
    if (this.sessions.has(sessionId)) return;
    this.sessions.set(sessionId, [...history]);
    this.touch(sessionId);
  }

  /** Prepend older history to an existing session's message list. */
  prependHistory(sessionId: string, olderMessages: ConversationMessage[]): void {
    const existing = this.sessions.get(sessionId);
    if (!existing) return;
    this.sessions.set(sessionId, [...olderMessages, ...existing]);
    this.touch(sessionId);
  }

  /**
   * Remove sessions that have not been accessed for `maxAgeSeconds`.
   * Default: 72 hours, matching the Python SDK.
   */
  cleanupExpired(maxAgeSeconds = 72 * 3600): void {
    const now = Date.now() / 1000;
    for (const [sid, ts] of this.lastAccess.entries()) {
      if (now - ts > maxAgeSeconds) {
        this.sessions.delete(sid);
        this.lastAccess.delete(sid);
      }
    }
  }

  // ── internal ─────────────────────────────────────────────────

  private ensureSession(sessionId: string): void {
    if (!this.sessions.has(sessionId)) {
      this.sessions.set(sessionId, []);
    }
    this.touch(sessionId);
  }

  private touch(sessionId: string): void {
    this.lastAccess.set(sessionId, Date.now() / 1000);
  }

  private trim(sessionId: string): void {
    const msgs = this.sessions.get(sessionId);
    if (!msgs) return;
    const cap = this.maxHistory * 2;
    if (msgs.length > cap) {
      this.sessions.set(sessionId, msgs.slice(-cap));
    }
  }
}
